library(ggplot2)
library(stringr)
library(plyr)
library(reshape2)

currentTheme = theme(strip.text.x = element_text(size = 18, angle = 90),
                     strip.text.y = element_text(size = 18, angle = 00),
                     axis.text.x = element_text(angle = 90, hjust = 1, size = 24),
                     axis.text.y = element_text(size = 24),
                     axis.title.x = element_text(size = 26),
                     axis.title.y = element_text(size = 26),
                     legend.text = element_text(size=22),
                     legend.position = "right",
                     legend.direction = "vertical") + theme_bw()

# Take a pileup file from GATK Pileup and turn it into a dataframe
readPileupFile <- function(pileupFile) {
  pileup = read.table(pileupFile, col.names = c("CHROM", "POS", "REF", "Pile", "QUAL"), sep = " ", quote = "", stringsAsFactors = FALSE)
  pileup = pileup[, c("CHROM", "POS", "REF", "Pile")]

  # Count the number of occurances of A, T, C, G,  and N
  pileup = ddply(pileup, names(pileup), function(x) {
    A = str_count(x$Pile, "A")
    T = str_count(x$Pile, "T")
    C = str_count(x$Pile, "C")
    G = str_count(x$Pile, "G")
    N = str_count(x$Pile, "N")
    depth = nchar(as.character(x$Pile))
    data.frame(A, T, C, G, N, depth)
  })
  
  pileup = pileup[, c("CHROM", "POS", "REF", "A", "T", "C", "G", "N", "depth")]
  pileup
}

# Plot results and save to pdf files
createAndSavePlots <- function(filePrefix, plotTitle, siteStatistics, sensitivityStatistics) {
  # Create a common color pallete
  truthStatusColors = c("False Positive" = "firebrick3", "True Positive" = "palegreen3", "Filtered" = "wheat3", "False Negative" = "mediumpurple3")
  
  # Plot sensitivity as a function of depth.
  binnedSensitivityPlot <- ggplot(sensitivityStatistics, aes(x = binnedDepth, y = Sensitivity, ymin = LowerSensitivity, ymax = UpperSensitivity)) + geom_errorbar() + 
    geom_point(mapping=aes(x=binnedDepth, y=Sensitivity), size=4, shape=21, fill="white") + ggtitle(paste0("Sensitivity vs depth ", plotTitle)) + currentTheme + ggsave(paste0(filePrefix, ".binnedSensitivity.pdf"))

  # Create histogram of TLOD scores that illustrates the distribution of TLOD scores for TP and FP calls
  siteStatistics$TLOD = as.numeric(as.character(siteStatistics$TLOD))
  tlodPlot = ggplot(siteStatistics, aes(TLOD, fill = CallStatus)) + geom_histogram(bins = 30, alpha = 0.8, color = "black") + xlim(6, 50) + 
    scale_fill_manual(values = truthStatusColors) + ggtitle(paste0("Histogram of tlod ", plotTitle)) + currentTheme + ggsave(paste0(filePrefix, ".tlod.pdf"))
  
  # Create histogram of Allele Fractions of various calls by TP, FP and FN.  
  afPlot = ggplot(siteStatistics, aes(alleleFraction, fill = CallStatus, color = "white")) + geom_histogram(bins = 30, alpha = 0.8, color = "black") + 
    scale_fill_manual(values = truthStatusColors) + xlim(0, 0.05) + ggtitle(paste0("Histogram of AF ", plotTitle)) + currentTheme + ggsave(paste0(filePrefix, ".af.pdf"))

  
  # Create histogram of depths for TP, FP, and FN calls.
  depthPlot = ggplot(siteStatistics, aes(depth, fill = CallStatus, color = "white")) + geom_histogram(bins = 30, alpha = 0.8, color = "black") + 
    scale_fill_manual(values = truthStatusColors) + ggtitle(paste0("Histogram of depths ", plotTitle)) + currentTheme + ggsave (paste0(filePrefix, ".depth.pdf"))
  
  # Create scatter plot of depth vs AltCount
  depthVsAltPlot = ggplot(siteStatistics, aes(x = depth, y = AltCount, color = CallStatus)) + geom_point(alpha = 0.8) + 
    ylim(0, max(siteStatistics$depth)*2*mean(siteStatistics$alleleFraction)) + scale_color_manual(values = truthStatusColors) + ggtitle(paste0("depth vs AltCount ", plotTitle)) + currentTheme + ggsave(paste0(filePrefix, ".depthVsAlt.pdf"))
  
  # Create boxplot of CallStatus vs AF
  callStatusVsAF = ggplot(siteStatistics, aes(x = CallStatus, y = alleleFraction)) + geom_boxplot() + ggtitle(paste0("CallStatus vs AF ", plotTitle)) + currentTheme + ggsave(paste0(filePrefix, ".callStatusVsAF.pdf"))
}

# evaluateTruth creates a data frame of truth statistics that are
# needed to generate plots for the report.
# truthTable is a table of truth variants
# evalTable is a table of variants generated by the pipeline
# pileupFile is a file generated from running GATK3 Pileup on the tumor bam.
evaluateTruth <- function(truthTable, evalTable, pileupFile) {
  truth = read.table(truthTable, header = TRUE, stringsAsFactors = FALSE)
  variants = read.table(evalTable, header = TRUE, stringsAsFactors = FALSE)
  pileup = readPileupFile(pileupFile)
  
  truth$truthStatus = "TrueVariant"
  truth$FILTER <- NULL

  if(nrow(variants) == 0) {
     variants$Called = character(0)
  } else {
     variants[, "Called"] = TRUE    
  }
  
  # Merge the called variants with the truth variants.
  siteLevelStatistics = merge(variants, truth, all=TRUE)
  siteLevelStatistics = ddply(siteLevelStatistics, names(siteLevelStatistics), function(x) {

    # If the sites filter status is NA, set to "No Call"
    if(is.na(x$FILTER)) {
      FILTER = "No Call"
    }
    else {
      FILTER = x$FILTER
    }
    
    # If the TLOD is NA, set it to 0
    if(is.na(x$TLOD)) {
      TLOD = 0
    }
    else {
      TLOD = x$TLOD
    }
    
    # If the truth status is NA, set it as a FalseVariant since
    # we don't see it in the truthset.
    if(is.na(x$truthStatus)) {
      truthStatus = "FalseVariant"
    }
    else {
      truthStatus = x$truthStatus
    }
    
    if(is.na(x$Called)) {
      Called = FALSE
    }
    else {
      Called = x$Called
    }
    data.frame(FILTER, TLOD, truthStatus, Called)
  })
  
  siteLevelStatistics = ddply(siteLevelStatistics, names(siteLevelStatistics), function(x) {
    CallStatus = "Unknown"
    if(x$truthStatus == "TrueVariant" & x$FILTER == "PASS") {
       CallStatus = "True Positive"
    }
    if(x$truthStatus == "FalseVariant" & x$FILTER == "PASS") {
      CallStatus = "False Positive"
    }
    if(x$truthStatus == "FalseVariant" & x$FILTER != "PASS") {
      CallStatus = "Filtered"
    }
    if(x$truthStatus == "TrueVariant" & x$FILTER != "PASS") {
      CallStatus = "False Negative"
    }
    
    FileName = gsub(".table", "", basename(evalTable))
    data.frame(CallStatus, FileName)
  })
  
  siteLevelStatistics = merge(siteLevelStatistics, pileup, all.x = TRUE)
  
  siteLevelStatistics = ddply(siteLevelStatistics, names(siteLevelStatistics), function(x) {
     RefCount = 0
     AltCount = 0
     ref = as.character(x$REF)
     alt = as.character(x$ALT)
     if(ref == "A") { RefCount = x[, "A"] }
     if(ref == "T") { RefCount = x[, "T"] }
     if(ref == "C") { RefCount = x[, "C"] }
     if(ref == "G") { RefCount = x[, "G"] }
     
     if(alt == "A") { AltCount = x[, "A"] }
     if(alt == "T") { AltCount = x[, "T"] }
     if(alt == "C") { AltCount = x[, "C"] }
     if(alt == "G") { AltCount = x[, "G"] }
     
     alleleFraction = AltCount / (RefCount + AltCount)
     data.frame(RefCount, AltCount, alleleFraction)
  })

  # bin the depth according to binWidth
  binWidth = 200
  siteLevelStatistics = ddply(siteLevelStatistics, names(siteLevelStatistics), function(x) {
    binnedDepth = floor((x$depth + binWidth/2) / binWidth)*binWidth
    data.frame(binnedDepth)
  })
  
  siteLevelStatistics
}

makeplot <- function(truthTable, evalTable, pileupFile, filePrefix) {
  confidenceLevel = 0.95
  
  resultsOutput = paste0(filePrefix, ".results.tsv")
  summaryResultsOutput = paste0(filePrefix, ".summary.results.tsv")

  siteLevelStatistics = evaluateTruth(truthTable, evalTable, pileupFile)
  
  sensitivityStatistics = count(siteLevelStatistics, c("CallStatus", "binnedDepth"))
  sensitivityStatistics = ddply(sensitivityStatistics, names(sensitivityStatistics), function(x) {
    TP = sensitivityStatistics[sensitivityStatistics$CallStatus == "True Positive" &
              as.character(sensitivityStatistics$binnedDepth) == as.character(x$binnedDepth), ]$freq
    
    FP = sensitivityStatistics[sensitivityStatistics$CallStatus == "False Positive" &
              as.character(sensitivityStatistics$binnedDepth) == as.character(x$binnedDepth), ]$freq
    
    FN = sensitivityStatistics[sensitivityStatistics$CallStatus == "False Negative" &
              as.character(sensitivityStatistics$binnedDepth) == as.character(x$binnedDepth) & !is.na(x$binnedDepth), ]$freq
    
    if(length(FN) == 0) {
      FN = 0
    }
    if(length(FP) == 0) {
      FP = 0
    }
    if(length(TP) == 0) {
      TP = 0
    }
    
    Sensitivity = TP / (TP + FN)
    
    # Get upper and lower bounds for sensitivity using the Clopper-Pearson method. 
    print(x)
    if(TP == 0 & FN == 0) {
      UpperSensitivity = 0
      LowerSensitivity = 0
    } else {
      UpperSensitivity = binom.test(x = TP, n = TP + FN, conf.level = confidenceLevel)$conf.int[2]
      LowerSensitivity = binom.test(x = TP, n = TP + FN, conf.level = confidenceLevel)$conf.int[1]
    }
    data.frame(Sensitivity, UpperSensitivity, LowerSensitivity, TP, FP, FN)
  })
  
  createAndSavePlots(filePrefix, filePrefix, siteLevelStatistics, sensitivityStatistics)

  # Write out table that will contain truth status for all called variants, and
  # variants that exist in the truthset, but were uncalled (false negatives).
  write.table(siteLevelStatistics, file = resultsOutput, quote = FALSE, sep = "\t", row.names = FALSE)
  
  TP_1000 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "True Positive" & siteLevelStatistics$depth > 1000, ])
  TP_1500 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "True Positive" & siteLevelStatistics$depth > 1500, ])
  TP_2000 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "True Positive" & siteLevelStatistics$depth > 2000, ])

  FN_1000 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "False Negative" & siteLevelStatistics$depth > 1000, ])
  FN_1500 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "False Negative" & siteLevelStatistics$depth > 1500, ])
  FN_2000 = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "False Negative" & siteLevelStatistics$depth > 2000, ])
  
  # Create summary statistics
  TP_total = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "True Positive", ])
  FP_total = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "False Positive", ])
  FN_total = nrow(siteLevelStatistics[siteLevelStatistics$CallStatus == "False Negative", ])
  Filtered_negatives = nrow(siteLevelStatistics[siteLevelStatistics$truthStatus == "FalseVariant" & siteLevelStatistics$FILTER != "No Call" & siteLevelStatistics$FILTER != "PASS", ])
  Filtered_positives = nrow(siteLevelStatistics[siteLevelStatistics$truthStatus == "TrueVariant" & siteLevelStatistics$FILTER != "No Call" & siteLevelStatistics$FILTER != "PASS", ])
  sensitivity_total = round(TP_total / (TP_total + FN_total), digits = 4)
  sensitivity_1000 = round(TP_1000 / (TP_1000 + FN_1000), digits = 4)
  sensitivity_1500 = round(TP_1500 / (TP_1500 + FN_1500), digits = 4)
  sensitivity_2000 = round(TP_2000 / (TP_2000 + FN_2000), digits = 4)
  
  # Calculate mean observed allele fraction over truth variants (TP and FN)
  # This gives an indication for what the true experimental spike in was.
  mean_af = round(mean(siteLevelStatistics[siteLevelStatistics$CallStatus == "True Positive" | siteLevelStatistics$CallStatus == "False Negative", ]$alleleFraction),
                       digits = 4)
  
  summary_results = melt(data.frame(TP_total, FP_total, FN_total, 
                                    sensitivity_total, sensitivity_1000, sensitivity_1500, sensitivity_2000, 
                                    Filtered_negatives, Filtered_positives,
                                    mean_af))
  summary_results$filePrefix = filePrefix
  
  # Make summary of results available in a file.
  write.table(summary_results, file = summaryResultsOutput, quote = FALSE, sep = "\t", row.names = FALSE)

  siteLevelStatistics
}


